### PR TITLE
[Android] Add basic silent download mode support

### PR DESCRIPTION
**Pull request by XiaosongWei**
Please find intent to implement @ https://lists.crosswalk-project.org/pipermail/crosswalk-dev/2015-October/003133.html

This PR contains the basic silent download mode implementation.

1) The runtime APK will be downloaded in the background silently without any user interaction.
2) The essential ingredients (libxwalkcore.so, classes.dex, icudtl.dat and xwalk.pak) will be extracted from the downloaded APK and saved to application's private storage.
3) Download mode will be working exclusively, that is, at the first launch, even if the runtime APK had been installed on the device, it still requests to download the dedicated runtime APK and only load the core libraries from that APK

BUG=XWALK-5121

### CODE REVIEW

**Comment by XiaosongWei** [comment]
@lincsoon @halton Please take a look, this PR contains only the basic implementation, more PRs containing security check, packaging with make_apk, LZMA decompress will be coming soon after this got merged.

**Comment by XiaosongWei** [comment]
@axinging

File: runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java [hunk 1]
```diff
@@ -237,6 +246,23 @@ private boolean findSharedCore() {
         return true;
     }
 
+    private boolean findDownloadedCore() {
+        String libDir = mWrapperContext.getDir(XWALK_CORE_LIB_DIR, Context.MODE_PRIVATE).toString();
+        String dexPath = libDir + File.separator + XWALK_CORE_CLASSES_DEX;
+        String dexOutputPath = mWrapperContext.getDir(OPTIMIZED_DEX_DIR, Context.MODE_PRIVATE).getAbsolutePath();
```
Comment (lincsoon) [inline runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java]: Line length exceeded

File: runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java [hunk 2]
```diff
@@ -29,6 +31,9 @@
     private static final String WRAPPER_PACKAGE = "org.xwalk.core";
     private static final String BRIDGE_PACKAGE = "org.xwalk.core.internal";
     private static final String TAG = "XWalkLib";
+    private static final String XWALK_CORE_LIB_DIR = "extracted_xwalkcore";
```
Comment (lincsoon) [inline runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java]: This name is ambiguous. Please use something like "XWALK_CORE_EXTRACTED_DIR" to indicate it's extracted explicitly

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java [hunk 3]
```diff
@@ -41,6 +44,7 @@
     private static boolean sLibraryLoaded = false;
     private static boolean sRunningOnIA = true;
     private static final String PRIVATE_DATA_DIRECTORY_SUFFIX = "xwalkcore";
+    private static final String XWALK_CORE_LIB_DIR = "extracted_xwalkcore";
```
Comment (lincsoon) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java]: This name is ambiguous

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java [hunk 4]
```diff
@@ -248,8 +248,14 @@ public InputStream openRawResource(String resource) {
                     String resourceName = resource.split("\\.")[0];
                     int resourceId = getResourceId(context, resourceName, "raw");
                     try {
-                        return context.getResources().openRawResource(resourceId);
-                    } catch (NotFoundException e) {
+                        try {
+                            return context.getResources().openRawResource(resourceId);
+                        } catch (NotFoundException e) {
+                            String resDir = context.getApplicationContext().getDir(
```
Comment (lincsoon) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java]: resDir could be a final variable outside this function

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java [hunk 5]
```diff
@@ -18,6 +20,7 @@
 import android.content.res.AssetManager;
 import android.content.res.Resources.NotFoundException;
 import android.os.Build;
+import android.os.Environment;
```
Comment (lincsoon) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java]: This is unused

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java [hunk 6]
```diff
@@ -248,8 +248,14 @@ public InputStream openRawResource(String resource) {
                     String resourceName = resource.split("\\.")[0];
                     int resourceId = getResourceId(context, resourceName, "raw");
                     try {
-                        return context.getResources().openRawResource(resourceId);
-                    } catch (NotFoundException e) {
+                        try {
+                            return context.getResources().openRawResource(resourceId);
+                        } catch (NotFoundException e) {
```
Comment (lincsoon) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java]: It's better to have a way to check if it's download mode like `isSharedMode`. But current implementation is okay

File: runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java [hunk 7]
```diff
@@ -151,7 +156,11 @@ public static int attachXWalkCore(Context context) {
         Log.d(TAG, "Attach xwalk core");
         sProvisionalInstance = new XWalkCoreWrapper(context, -1);
         if (!sProvisionalInstance.findEmbeddedCore()) {
-            sProvisionalInstance.findSharedCore();
+            if (XWalkUpdater.isDownloadModeEnabled()) {
```
Comment (lincsoon) [inline runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java]: The second startup of the app won't trigger the XWalkUpdater. So this way of checking download mode is incomplete

**Comment by XiaosongWei** [comment]
@lincsoon Revised, please take a look.

**Comment by XiaosongWei** [comment]
Ping... @lincsoon

File: runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java [hunk 8]
```diff
@@ -237,6 +248,39 @@ private boolean findSharedCore() {
         return true;
     }
 
+    private boolean findDownloadedCore() {
+        String libDir = mWrapperContext.getDir(XWALK_CORE_EXTRACTED_DIR, Context.MODE_PRIVATE).
+                toString();
```
Comment (lincsoon) [inline runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java]: Should be `getAbsolutePath()`?

File: runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java [hunk 9]
```diff
@@ -237,6 +248,39 @@ private boolean findSharedCore() {
...
+        mCoreStatus = XWalkLibraryInterface.STATUS_MATCH;
+        return true;
+    }
+
+    private boolean isDownloadMode() {
+        try {
+            PackageManager packageManager = mWrapperContext.getPackageManager();
+            ApplicationInfo appInfo = packageManager.getApplicationInfo(
+                    mWrapperContext.getPackageName(), PackageManager.GET_META_DATA);
+            final String enableStr = appInfo.metaData.getString(META_XWALK_ENABLE_DOWNLOAD_MODE);
```
Comment (lincsoon) [inline runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java]: Why use `final`?

File: runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java [hunk 10]
```diff
@@ -237,6 +248,39 @@ private boolean findSharedCore() {
...
+            if (enableStr != null && enableStr.equalsIgnoreCase("enable")) {
```
Comment (lincsoon) [inline runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java]: Just `return enableStr.equalsIgnoreCase("enable")` is okay. The `NullPointerException` will be caught.

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java [hunk 11]
```diff
@@ -189,6 +196,20 @@ public void run() {
         });
     }
 
+    private static boolean isDownloadMode(final Context context) {
+        try {
+            PackageManager packageManager = context.getPackageManager();
+            ApplicationInfo appInfo = packageManager.getApplicationInfo(
+                    context.getPackageName(), PackageManager.GET_META_DATA);
+            final String enableStr = appInfo.metaData.getString(META_XWALK_ENABLE_DOWNLOAD_MODE);
+            if (enableStr != null && enableStr.equalsIgnoreCase("enable")) {
```
Comment (lincsoon) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java]: Ditto

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java [hunk 12]
```diff
@@ -244,6 +261,14 @@ public InputStream openRawResource(String resource) {
                     } catch (IOException e) {
                         Assert.fail(resource + " can't be found in assets.");
                     }
+                } else if (isDownloadMode(context)) {
```
Comment (lincsoon) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java]: It's not good to check download mode from the manifest every time

Comment (lincsoon) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java]: It's not necessary to check download mode from the manifest every time

**Comment by lincsoon** [comment]
LGTM

**Comment by rakuco** [comment]
Weirdly enough, this broke the build from scratch here, but not our canaries. I've sent #3355 to fix it.

**Pull request merged**